support-frontend/assets/helpers/tracking/quantumMetric.ts

```diff
@@ -0,0 +1,70 @@
+import { getGlobal, isSwitchOn } from 'helpers/globalsAndSwitches/globals';
+import { logException } from 'helpers/utilities/logger';
+
+const pathIsValid = (): boolean => {
+	const locationPath = window.location.pathname;
+	// Contribution landing page and thank you page (all regions)
+	const contributionsRegex = /\/contribute(\/thankyou)?(\?.*)?$/;
+	// Non-Gifting Digi Subs landing page, checkout and thank you page (all regions)
+	const digiSubRegex = /\/subscribe\/digital(\/checkout|\/thankyou)?(\?.*)?$/;
+
+	return [contributionsRegex, digiSubRegex].some(
+		(pathRegEx) => locationPath.match(pathRegEx) != null,
+	);
+};
+
+const addQM = (): void => {
+	const qtm = document.createElement('script');
+	qtm.type = 'text/javascript';
+	qtm.async = true;
+	qtm.src =
+		'https://cdn.quantummetric.com/instrumentation/1.31.3/quantum-gnm.js';
+	qtm.integrity = 'sha256-EgTDwEe1o6/yOkCHRanpJrDBbdGyHnc+mxy6Cxd9vzk=';
+	qtm.crossOrigin = 'anonymous';
+	const d = document.getElementsByTagName('script')[0];
+	if (!window.QuantumMetricAPI && d.parentNode) {
+		d.parentNode.insertBefore(qtm, d);
+	}
+};
+
+const init = (): void => {
+	/**
+	 * Return immediately if we're Server Side Rendering OR
+	 * The feature switch is off OR
+	 * the window.location.pathname is NOT valid.
+	 *
+	 */
+	if (
+		getGlobal('ssr') ||
+		!isSwitchOn('enableQuantumMetric') ||
+		!pathIsValid()
+	) {
+		return;
+	}
+
+	/**
+	 * Dynamically load @guardian/consent-management-platform
+	 * on condition we're not server side rendering (ssr) the page.
+	 * @guardian/consent-management-platform breaks ssr otherwise.
+	 */
+	import('@guardian/consent-management-platform')
+		.then(({ onConsentChange }) => {
+			onConsentChange((state) => {
+				// Check users consent state
+				if (
+					state.ccpa?.doNotSell === false ||
```

Comment:
Slightly struggling with the logic here, so if `doNotSell` is true we DO load QM? And also if the three consents are true.

Reply:
Yeah the various potential return values for consent can make things confusing, they are explained here: https://github.com/guardian/consent-management-platform.

We only want to load the script if the condition on the lefthand side OR righthand side of the double pipe operator (`||`) is met.

We first check the conditional on the lefthand side. `state.ccpa` should only be returned for users in the USA. `state.ccpa.doNotSell` reports whether the user has withdrawn consent to sell their data in the USA, so a `true` value indicates they have withdrawn consent and a `false` value indicates they have NOT withdrawn consent. We only want to load the script for user's in the USA who have NOT withdrawn consent. 

If the condition on the lefthand side isn't met we then check the conditional on the righthand side. `state.tcfv2` should be defined for non-US users. `state.tcfv2.consents` reports the user's preferences for each of the TCFv2 purposes, a `true` value indicates they have consented to a purpose, we want to check that they've consented to the 3 purposes listed.